[IMP] base: Allow rendering barcodes as SVG

ivantodorovich opened the pull request:
**Description of the issue/feature this PR addresses:**

Introduce a new `format` parameter which can be used to choose among 'png' (default) and 'svg' for the barcode rendering.

SVGs have unlimited scaling capabilities, which is great when printing really small or really big barcodes in labels, etc; as there's absolutely no quality loss when resizing


---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/108770).

ivantodorovich commented on odoo/addons/base/models/ir_actions_report.py:
```diff
@@ -564,7 +565,11 @@ def barcode(self, barcode_type, value, **kwargs):
-            return barcode.asString('png')
+            res = barcode.asString(kwargs['format'])
+            if kwargs['format'] == 'svg' and isinstance(res, str):
+                res = res.encode('utf-8')
```
To reviewers: the reportlib's `asString` method returns `bytes` for raster images and `str` for svgs
This is here for two reasons: a) to be consistent and b) to be able to use `guess_mimetype` on the result

ivantodorovich commented on odoo/addons/base/models/ir_actions_report.py:
```diff
@@ -554,7 +555,7 @@ def barcode(self, barcode_type, value, **kwargs):
 
         try:
-            barcode = createBarcodeDrawing(barcode_type, value=value, format='png', **kwargs)
+            barcode = createBarcodeDrawing(barcode_type, value=value, **kwargs)
```
To reviewers: the `format='png'` kwarg isn't used by `createBarcodeDrawing`. It's useless to have it here.
It's here since many years ago, when `createBarcodeImageInMemory` was used instead.

Julien00859 commented on odoo/http.py:
```diff
@@ -1938,7 +1938,7 @@ def set_csp(self, response):
 
         mime, _params = cgi.parse_header(headers.get('Content-Type', ''))
-        if not mime.startswith('image/'):
+        if mime == 'image/svg+xml' or not mime.startswith('image/'):
```
This would disable CSP for svg images, why would you need to disable CSP?

ivantodorovich replied:
Unlike other images, svgs are xml documents and their elements may contain styles.

Without this change, the browser renders a malformed image due to inline styles not being applied. (for instance, a barcode renders as a black square).

I'm not too familiar with CSP, though.
Maybe there's a way to bee a little bit more permisive with svgs, without fully disabling it?

Julien00859 replied:
Indeed, that's what CSP is for, to forbid content injection from untrusted sources. Here we are enabling svg images only for barcodes which I suppose would be self-contained and plain-dark. Can you provide an example where disallowing inline-css on svg images makes it hard to read a barcode?

Julien00859 replied:
Looking at the blog post you sent, maybe we should indeed craft a specific CSP for SVG

ivantodorovich replied:
Take for example the QR code generated from this URL (browser: Google Chrome):

`http://localhost:8069/report/barcode/QR/asd?format=svg&height=100&width=100`

**With the generic CSP rule, we get this:**
![image](https://user-images.githubusercontent.com/1914185/210274387-c50b9152-dc65-4c04-8cb6-d4a5492fa7dd.png)

**With the new svg-specific CSP rule (or none at all):**
![image](https://user-images.githubusercontent.com/1914185/210274344-f47cea98-a634-4e1e-8e44-b31cd7e16f91.png)

**By inspecting the svg element we can see the issue is some inline styles not being applied:**
<details>
<summary>See screenshot</summary>

![image](https://user-images.githubusercontent.com/1914185/210274459-bb6057ea-98ad-4166-92df-da6acc43d958.png)
</details>

---

Finally, and following some tips found online (that blog post included), I added the `style-src 'self' 'unsafe-inline'` rule only for `image/svg+xml`, and it's working nicely (at least for barcodes generated by reportlib)

mart-e replied:
Is the unsafe-inline still required? [The bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1262842) seems solved with only `style-src 'self'`

ivantodorovich replied:
Actually, AFAICS that bug report is about Firefox incorrectly applying the CSP policy when rendering `svg` on an `img` tag (through the `src` attribute)

You see, rendering images through the `img` tag (`<img src="..." />`) is considered safe by the browsers, and thus the page's CSP is completely ignored (apart from the `image-src` rule, that must be valid) as it's deemed not necessary.

Apparently, Firefox had a bug where the SVG's inline styles were silently ignored when rendered using an `img` tag. That is solved, but it's not really our issue here.

We do need `'unsafe-inline'`, because we want styles that are inlined in the svg to be applied. There isn't any category/rule for "safe" styles, that rules blocks all inline styles. Although, in reality, they aren't really that unsafe (see for example https://github.com/angular/angular/issues/6361).

---

See my comment below for more

Julien00859 reviewed (changes requested):
Can you refine `guess_mimetype` to only allow png and svg?

Julien00859 commented:
Don't forget:
> Can you refine guess_mimetype to only allow png and svg?

ivantodorovich commented:
> Can you refine guess_mimetype to only allow png and svg?

How about like this? https://github.com/odoo/odoo/pull/108770/commits/7b9043ee2c1a713ee6025fad77494bea0e036a73

So we keep the image format responsibility in the `ir.actions.report` model and out of the controller (as long as it's an image)

Julien00859 commented:
Ok there was something I didn't understand. So the `format` argument was hardcoded before and now it can be part of the `kwargs`. We should change the docstring of the route to document this `format` kwarg. I didn't see that that format was restricted to png and svg thanks to the lambda so I guess it is useless to assert the guess_mimetype return value.

ivantodorovich commented:
> Ok there was something I didn't understand. So the format argument was hardcoded before and now it can be part of the kwargs. We should change the docstring of the route to document this format kwarg. I didn't see that that format was restricted to png and svg thanks to the lambda so I guess it is useless to assert the guess_mimetype return value.

Ohh, gotcha.

This should be ready now, then ;) 
I added the docstring and squashed.

Julien00859 commented on addons/web/controllers/report.py:
```diff
@@ -69,6 +70,7 @@ def report_barcode(self, barcode_type, value, **kwargs):
         'Standard93', 'UPCA', 'USPS_4State'
         :param width: Pixel width of the barcode
         :param height: Pixel height of the barcode
+        :param foramt: Accepted values: 'png' (default) or 'svg'.
```
foramt => format at least
```suggestion
        :param format: format of the generated image, 'png' (default) or 'svg'
```

ivantodorovich replied:
🙈 done

Julien00859 reviewed (approved):
I'm ok implementation-wise, now let's wait for the stock team to agree feature-wise and the security team to agree for the CSP

ivantodorovich commented:
> I'm ok implementation-wise, now let's wait for the stock team to agree feature-wise and the security team to agree for the CSP

Thanks a lot  !

mart-e commented:
Could we maybe move the CSP change at the controller level instead of globally? I am not very fan of adding a `unsafe` rule in CSP, we never know how it will be used by other modules.

ivantodorovich commented:
> Could we maybe move the CSP change at the controller level instead of globally? I am not very fan of adding a unsafe rule in CSP, we never know how it will be used by other modules.

Indeed. I've just done that to limit the scope of this PR.
However, please read the following 🙏🏻 

---

I didn't really know much about CSP before this, but after doing some reading and looking at how it's implemented within Odoo, I think the issue is bigger than just which rule we apply for svg's.

CSP allows us to define some rules on html _or xml_ documents for the browser to know which of its resources to trust and thus load/process. The objective is to mitigate XSS attacks.

The rules are defined for html documents, not its resources. And this is not what Odoo is doing. [Odoo is only setting the csp headers for `image/*` mimetypes](https://github.com/odoo/odoo/blob/964dabd8d9d11e7ce8205b2eb0a2e6ee83c9f635/odoo/http.py#L2043-L2044) (with the exception of some specific controllers that I'll mention below, but not relevant to this point)

Having the CSP on a `png` or `jpg` file server response is the less useful scenario, certainly not the real reason CSP exists. As CSP applies on the html document, having them on image responses makes the rule apply only on the auto-generated HTML the browsers display when accessing the image directly through its URL. It only protects against attackers somehow being able to modify this auto-generated HTML, and only when the user accesses the image URL directly.

It's certainly important to have, and specially more important on SVGs (as it's treated as an html document for this matter), but it is not the most important use of CSP as, again, it only applies for the case when the user accesses the image URL directly.

In the case of images within an HTML (the most common scenario), the CSP defined in **their** response headers are **not** used when the image is used in the `src` attribute of an `img` element. This is considered safe by the browser, and it has some restrictions for SVGs (no javascript, no external images, etc..). On these cases, it's the CSP defined on the HTML document that applies (being able to load the image or not, depending on the HTML document's image-src policy, for example). The only cases where the CSP defined on the image response is used is when the image is embedded using `iframe`, `embed` or, in case of svg, `svg` elements.. and in those cases these elements are already sandboxed by the browser, and the CSP applies for it's own sandboxed environment, not the parent HTML document.

So.. The most important take-off of this analysis is that, in fact, Odoo doesn't really have CSP enabled. Not for it's HTML documents, which is the whole point. I think this topic should be deeply revisited if you consider CSP to be important.

---

Regarding my comment about some controllers having specific CSP rules, they are these:
- https://github.com/odoo/odoo/blob/5b26a080ae28fab2a78e92a838472578591eef17/addons/auth_signup/controllers/main.py#L71
- https://github.com/odoo/odoo/blob/5b26a080ae28fab2a78e92a838472578591eef17/addons/auth_signup/controllers/main.py#L109
- https://github.com/odoo/odoo/blob/5b26a080ae28fab2a78e92a838472578591eef17/addons/portal/controllers/portal.py#L218

However they all seem to be setting only the `frame-ancestors` rule, and a warning ⚠️ : this pattern for setting controller-specific CSP rules [will completely overwrite the rules set by `set_csp`](https://github.com/odoo/odoo/blob/964dabd8d9d11e7ce8205b2eb0a2e6ee83c9f635/odoo/http.py#L2039-L2040) (if there were any for html documents!)

ivantodorovich commented:
Hi @mart-e ! I've applied the change in order to set the csp rule to the controller level.

Please let me know if there's anything else I could do to speed this up ;)

ivantodorovich commented:
* PR Rebased ✅

ivantodorovich commented:
PR rebased

Is there hope to see this merged before the next release? 😅
